### PR TITLE
[bgpcfgd] ECMP overlay VxLan with BGP support

### PR DESCRIPTION
**Pull request by StormLiangMS**
#### Why I did it
https://github.com/Azure/SONiC/blob/master/doc/vxlan/Overlay%20ECMP%20with%20BFD.md
From the design, need to advertise the route with community string, the PR is to implement this.

#### How I did it
To use the route-map as the profile for the community string, all advertised routes can be associated with one route-map. 
Add one file, mangers_rm.py, which is to add/update/del the route-map. Modified the managers_advertise_rt.py file to associate profile with IP route.

The route-map usage is very flexible, by this PR, we only support one fixed usage to add community string for route to simplify this design.

#### How to verify it
Implement new unit tests for mangers_rm.py and updated unit test for managers_advertise_rt.py.
Manually verified the test case in the test plan section, will add testcase in sonic-mgmt later. https://github.com/Azure/sonic-mgmt/issues/5581

### 2.8.3 BGP advertising
The below cases are executed first for IPv4 and repeat the same for IPv6. 
| Step | Goal | Expected results |
|-|-|-|
| Create a tunnel route and advertise the tunnel route to all neighbor without community id | BGP | ALL BGP neighbors can recieve the advertised BGP routes |
| Create a tunnel route and advertise the tunnel route to all neighbor with community id | BGP | ALL BGP neighbors can recieve the advertised BGP routes with community id |
| Update a tunnel route and advertise the tunnel route to all neighbor with new community id | BGP | ALL BGP neighbors can recieve the advertised BGP routes with new community id |
| Create a tunnel route and advertise the tunnel route to all neighbor with BGP profile, but create the profile later| BGP | ALL BGP neighbors can recieve the advertised BGP routes without community id first, after the profile table created, the community id would be added and all BGP neighbors can recieve this update and associate the community id with the route |
| Delete a tunnel route | BGP | ALL BGP neighbors can remove the previously advertised BGP routes |
| Create 4k tunnel routes and advertise all tunnel routes to all neighbor with community id | BGP scale | ALL BGP neighbors can recieve 4k advertised BGP routes with community id and record the time |
| Updat BGP_PROFILE_TABLE with new community id for 4k tunnel routes and advertise all tunnel routes to all neighbor with new community id | BGP scale | ALL BGP neighbors can recieve 4k advertised BGP routes with new community id and record the time |

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py [hunk 1]
```diff
@@ -18,82 +22,140 @@ def __init__(self, common_objs, db, table):
...
+                log_err("BGPAdvertiseRouteMgr:: No valid ip %s for advertised route %s" % (ip_prefix[0], key))
+                return False
+        else:
+            return False
+        return True
+
+    def _set_handler_validate(self, key, data):
+        if data:
+            if ("profile" in data and data["profile"] in ROUTE_MAPS) or data == {"":""}:
+                return self._ip_addr_validate(key)
```
Comment (prsunny) [inline src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py:82]: I think, this could be just returning 'true' here. IP prefix should already be validated, and we may just be adding extra delay here.

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py [hunk 2]
```diff
@@ -18,82 +22,140 @@ def __init__(self, common_objs, db, table):
...
+            
+        log_err("BGPAdvertiseRouteMgr:: Invalid data %s for advertised route %s" % (data, key))
+        return False
 
-    def add_route_advertisement(self, vrf, ip_prefix):
+    def _del_handler_validate(self, key):
```
Comment (prsunny) [inline src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py:87]: This can be removed as well

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py [hunk 3]
```diff
@@ -18,82 +22,140 @@ def __init__(self, common_objs, db, table):
...
             cmd_list.append("router bgp %s vrf %s" % (bgp_asn, vrf))
 
         cmd_list.append(" address-family %s unicast" % ("ipv6" if is_ipv6 else "ipv4"))
-        cmd_list.append("  %snetwork %s" % ('no ' if op == self.OP_DELETE else '', ip_prefix))
 
-        self.cfg_mgr.push_list(cmd_list)
+        if data and "profile" in data:
+            cmd_list.append("  network %s route-map %s" % (ip_prefix, "%s_RM" % data["profile"]))
+            log_info(
+                "BGPAdvertiseRouteMgr:: Update bgp %s network %s with route-map %s"
```
Comment (prsunny) [inline src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py:127]: Based on scale, there could be significant number of route updates. So please reduce the log-level and ensure it won't show up in regular flow unless explicitly enabled

File: src/sonic-bgpcfgd/bgpcfgd/managers_rm.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            return
+        self._remove_rm(key)
+
+    def _remove_rm(self, rm):
+        cmds = ["no route-map %s permit 100" % ("%s_RM" % rm)]
+        log_info("BGPRouteMapMgr:: remove route-map %s" % ("%s_RM" % rm))
+        self.cfg_mgr.push_list(cmds)
+        log_info("BGPRouteMapMgr::Done")
+
+    def _set_handler_validate(self, key, data):
```
Comment (prsunny) [inline src/sonic-bgpcfgd/bgpcfgd/managers_rm.py:45]: Lets use consistent formatting for names. Can you check if its "__" vs "_" ?

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py [hunk 5]
```diff
@@ -18,82 +22,140 @@ def __init__(self, common_objs, db, table):
...
             table,
         )
-        
-        self.directory.subscribe([("CONFIG_DB", swsscommon.CFG_DEVICE_METADATA_TABLE_NAME, "localhost/bgp_asn"),], self.on_bgp_asn_change)
-        self.advertised_routes = dict()
-
 
-    OP_DELETE = 'DELETE'
-    OP_ADD = 'ADD'
+        self.directory.subscribe(
```
[the remark this answers is not in the document]
Comment (StormLiangMS) [reply src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py]: Sure, will keep original code as much as possible. Only change is to use " instead of ' to keep the style consistent.

File: src/sonic-bgpcfgd/bgpcfgd/main.py [hunk 6]
```diff
@@ -62,6 +63,8 @@ def do_work():
         StaticRouteMgr(common_objs, "CONFIG_DB", "STATIC_ROUTE"),
         # Route Advertisement Managers
         AdvertiseRouteMgr(common_objs, "STATE_DB", swsscommon.STATE_ADVERTISE_NETWORK_TABLE_NAME),
+        # TBD, Change to schema name from swsscommon (swsscommon.APP_BGP_PROFILE_TABLE_NAME) when the submodule is advanced.
```
[the remark this answers is not in the document]
Comment (StormLiangMS) [reply src/sonic-bgpcfgd/bgpcfgd/main.py]: I checked, the code is already advanced in the master branch. Will make sure which is advanced before cherry pick to 202012 branch.

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py [hunk 7]
```diff
@@ -18,82 +22,140 @@ def __init__(self, common_objs, db, table):
...
 
-
     def del_handler(self, key):
+        log_info("AdvertiseRouteMgr:: del handler")
+        if not self._del_handler_validate(key):
+            return
         vrf, ip_prefix = self.split_key(key)
         self.remove_route_advertisement(vrf, ip_prefix)
 
+    def _ip_addr_validate(self, key):
```
[the remark this answers is not in the document]
Comment (StormLiangMS) [reply src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py]: Good one, add one comment in the code.

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py (hunk 1 above)
Comment (StormLiangMS) [reply src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py]: Changed.

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py (hunk 2 above)
Comment (StormLiangMS) [reply src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py]: Removed

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py (hunk 3 above)
Comment (StormLiangMS) [reply src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py:94]: Good one, change to debug.

File: src/sonic-bgpcfgd/bgpcfgd/managers_rm.py (hunk 4 above)
Comment (StormLiangMS) [reply src/sonic-bgpcfgd/bgpcfgd/managers_rm.py]: Sure, changed all internal function to "__"

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py [hunk 8]
```diff
@@ -18,82 +22,105 @@ def __init__(self, common_objs, db, table):
...
-
     def del_handler(self, key):
+        log_debug("AdvertiseRouteMgr:: del handler")
         vrf, ip_prefix = self.split_key(key)
         self.remove_route_advertisement(vrf, ip_prefix)
 
-
-    def add_route_advertisement(self, vrf, ip_prefix):
+    def __set_handler_validate(self, key, data):
+        if data:
```
Comment (prsunny) [inline src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py:48]: If data is not present, we need to advertise without a community string (for backward compatibility). Can you please confirm.?

File: src/sonic-bgpcfgd/tests/test_advertise_rt.py [hunk 9]
```diff
@@ -62,7 +62,7 @@ def test_set_del():
     set_del_test(
         mgr,
         "SET",
-        ("fc00:10::/64", {}),
```
Comment (prsunny) [inline src/sonic-bgpcfgd/tests/test_advertise_rt.py:65]: Can we retain this case to cover empty data?

**Review by prsunny — COMMENTED** [review]
lgtm, minor comments

File: src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py (hunk 8 above)
Comment (StormLiangMS) [reply src/sonic-bgpcfgd/bgpcfgd/managers_advertise_rt.py:48]: @prsunny confirmed, the check after the 'or' is for the case which doesn't have the community string. 
if ("profile" in data and data["profile"] in ROUTE_MAPS) or data == {"":""}:

File: src/sonic-bgpcfgd/tests/test_advertise_rt.py (hunk 9 above)
Comment (StormLiangMS) [reply src/sonic-bgpcfgd/tests/test_advertise_rt.py:65]: @prsunny  I think the old unit test is inaccurate, from existing code of VNetRouteOrch, it send out {key:{"":""}} to the bgpcfgd for empty case as below. And also by manually testing,  can't insert a key with a "empty" to the redis.
void VNetRouteOrch::addRouteAdvertisement(IpPrefix& ipPrefix)
{
    const string key = ipPrefix.to_string();
    vector<FieldValueTuple> fvs;
    fvs.push_back(FieldValueTuple("", ""));
    state_vnet_rt_adv_table_->set(key, fvs);
}

**Review by prsunny — APPROVED** [review]
